core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4769,19 +4778,51 @@ private void convertValuesImpl(
       return;
     }
 
-    for (SqlNode rowConstructor1 : values.getOperandList()) {
-      SqlCall rowConstructor = (SqlCall) rowConstructor1;
+    SqlCall insertOp = callStack.peek();
+    boolean processDefaults = insertOp != null && containsDefault(values);
+
+    if (targetRowType == null) {
+      RelDataType listType = validator().getValidatedNodeType(values);
+      targetRowType = SqlTypeUtil.promoteToRowType(typeFactory, listType, null);
+    }
+
+    assert insertOp instanceof SqlInsert || !processDefaults : "operation: " + insertOp;
+
+    @Nullable InitializerExpressionFactory initializerFactory = null;
+    @Nullable RelOptTable targetTable = null;
+    int[] mapping = null;
+
+    if (processDefaults) {
+      targetTable = getTargetTable(requireNonNull(insertOp, "insertOp"));
+      initializerFactory =
+          getInitializerFactory(getNamespace(requireNonNull(insertOp, "insertOp"))
+              .getTable());
+      mapping =
+          processColumnsMapping(requireNonNull(targetRowType, "targetRowType"), targetTable);
+    }
+
+    for (SqlNode rowConstructor : values.getOperandList()) {
+      SqlCall rowConstructor0 = (SqlCall) rowConstructor;
```

Comment:
Does 0 mean anything else？If not,maybe remove 0 could make code clean.

Reply:
plz suggest - what form do you prefer ? 
```
SqlCall rowConst = (SqlCall) rowConstructor;
SqlCall newRowConst = (SqlCall) rowConstructor;
```
as for me - no one of them not looks good.
i think usage of "0" is clear, it suggest other engineers that this is local scoped param, no ?

Reply:
Hi @zstan ,for me, when I see a 0 or 1 in the code, I usually assume that the 0 or 1 has a special meaning, it may represent the element index information, or in some cases the first or second element.

If I were to suggest, I would suggest the second,but this is just my advice and thoughts,It has no impact on this PR.

So It's ok for you to follow your ideas:)

Reply:
done, thnks !